INTEGRATION.md

```diff
@@ -0,0 +1,152 @@
+
+The decision tree is influences by a couple environement variables.
+
+**`IPFS_RACING_GATEWAY`** When this environemnt variable isn't found (the default), racing gateways should be attempted upon reaching this point in the flow. When the environment variable exists it's value should be used. If it's value is `1` (or `true`) then racing gateways should be attempted. If this value is `0` (or `false`) then racing gateways are off. The control flow proceeds in the `no` branch.
+
+**`IPFS_FALLBACK_GATEWAY`** When this variable doesn't exist `dweb.link` will be used. When the variable does exist it's value will be used instead.
+
+### Gateway from command argument
+
+**This feature is optional and only for applications integrating IPFS support.**
+
+An application can opt to support a command line option to provide a gateway. If a user does provide this option then it should overrule any other gateway detection and be used as the gateway of choice. If implemented, it's recommended to go for either a `--gateway` or `--ipfs-gateway` argument. It depends very much on the application itself as to which option is most sensible.
+
+An example implementation that is doing this is ffmpeg with the ffplay utility. It allows the `-gateway` argument which by default is empty but can be set like: `-gateway http://127.0.0.1:8080` and would then be used to handle `ipfs://<cid>` or `ipns://<cid>`.
+
+### Gateway from IPFS_GATEWAY environment variable
+
+When there is no command line argument, the `IPFS_GATEWAY` environment is next. If it contains a value, it will be used as gateway.
+
+### Gateway file
+
+**This feature is only for IPFS implementers, not for IPFS application integrations.**
+
+When the implementation provides a gateway (and it's not disabled through other means) then it should make that known to other local applications. The gateway file serves this purpose. It's a file with only 1 single line containing the full http URL to your gateway. For example, it could contain the line: "http://localhost:8080".
+
+The file conditions:
+    1. is named "gateway"
+    2. **only** exists when your implementation actually starts a gateway
+    3. is removed when your implementation shuts down
+
+For historical and compatibility readons, this file shall be placed in:
+$HOME/.ipfs/ thus the resulting end path is going to be $HOME/.ipfs/gateway
+
+Future implementers must follow the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) where the gateway file will placed in:
+**$XDG_CONFIG_HOME/ipfs/gateway**
+
+The conditions for this file are the same as those in $HOME/.ipfs/gateway.
+
+In the, admitedly rare, event of running multiple IPFS implementations each hosting their own gateway. First-come-first-serve applies here. The application that created the gateway file owns it and takes care of removing it. Subsequent instances or different application should not touch the file if it's already there.
```

Comment:
> Subsequent instances or different application should not touch the file if it's already there.

How would this work? What locking mechanism you have in mind?
If I have `gateway` manually created to have `https://dweb.link`  and I start local IPFS node, it  will not be able to update the file? 

This feels like bad UX/DX, giving me more confidence that 
- this file should be a list of URLs (one per line)
- starting  new IPFS node on port `8081` should add localhost URL at the top of the file, but after any already existing localhost nodes:
  ```
  http://localhost:8080
  http://localhost:8081
  https://dweb.link
  ```
  - in the example above, `8080` was already running when `8081` started – this ensures consistent UX – node that started first will remain default local gateway until it exits – starting other localhost nodes will not suddenly hijack requests (it would be bad, especially when running tests and forgetting to adjust HOME dir)